### PR TITLE
FOGL-5665 Addition of new datapoint types and generic classes to handle transforming readings to Python

### PR DESCRIPTION
**Pull request by MarkRiddoch**
Eventually the idea is to replace all the implementation s for filters, south and north plugins to use the PythonReading and PythonReadingSet classes rather than have the duplication that we currently have. We also should look at the likes of the Python35 filter plugin, rule plugin, etc.

We have two new datapoint types, a data buffer, which is transformed to a numpy array in Python and a simple image format, again gets passed to Python as a two dimensional bumpy array. In JSON they are stored as base64 encoded data, so are not suitable for very large objects. The aim of them is to allow image and radiometric data to be passed into ML models written in Python.

### CODE REVIEW

File: docs/OMF.rst [hunk 1]
```diff
@@ -164,7 +172,11 @@ The second screen will request the following information:
    - **Number Format:** Used to match Fledge data types to the data type configured in PI. The defaults is float64 but may be set to any OMF datatype that supports floating point values.
    - **Compression:** Compress the readings data before sending it to the PI System.
 
+<<<<<<< HEAD
```
Comment (AmandeepArora) [inline docs/OMF.rst:175]: Conflicts needs to be resolved in this file.

Comment (MarkRiddoch) [reply docs/OMF.rst]: Thanks had not noticed that.

**Review by pintomax — DISMISSED** [review]

**Review by AmandeepArora — DISMISSED** [review]

**Review by pintomax — APPROVED** [review]

**Review by pintomax — APPROVED** [review]

**Review by pintomax — APPROVED** [review]

**Review by pintomax — APPROVED** [review]

**Review by pintomax — APPROVED** [review]

**Review by pintomax — APPROVED** [review]

File: C/services/south-plugin-interfaces/python/python_plugin_interface.cpp [hunk 2]
```diff
@@ -79,6 +79,13 @@ void *PluginInterfaceInit(const char *pluginName, const char * pluginPathName)
     PyList_Append(sysPath, PyUnicode_FromString((char *) southRootPath.c_str()));
     PyList_Append(sysPath, PyUnicode_FromString((char *) fledgePythonDir.c_str()));
 
+    // Set sys.argv for embedded Python 3.5
```
Comment (pintomax) [inline C/services/south-plugin-interfaces/python/python_plugin_interface.cpp:82]: Don't we need same fix for

C/services/notification-plugin-interfaces/python/python_plugin_interface.cpp ?

if yes, please add it

**Review by pintomax — CHANGES_REQUESTED** [review]

File: C/services/south-plugin-interfaces/python/python_plugin_interface.cpp (hunk 2 above)
Comment (MarkRiddoch) [reply C/services/south-plugin-interfaces/python/python_plugin_interface.cpp:82]: @AmandeepArora I am assuming the answer to this is yes we should add this for the notifications, correct?

Comment (AmandeepArora) [reply C/services/south-plugin-interfaces/python/python_plugin_interface.cpp:82]: Yes, adding.

Comment (AmandeepArora) [reply C/services/south-plugin-interfaces/python/python_plugin_interface.cpp:82]: Done, please verify using some notification python plugin by importing tensorflow.

Comment (pintomax) [reply C/services/south-plugin-interfaces/python/python_plugin_interface.cpp:82]: @AmandeepArora please suggest something easy to test. Some startup example code.
I.e. plugin skeleton, instructions for installing tensorflow etc

I just have the notification plugins with Numpy and nothing else.

Comment (AmandeepArora) [reply C/services/south-plugin-interfaces/python/python_plugin_interface.cpp:82]: You can install tensorflow using "pip install tensorflow"
And in python plugin, at top you can add "import tensorflow as tf" and in plugin_init, you can print "tf.__version__" using _LOGGER.info. That should suffice, earlier we were getting crash dump just at import time.

Comment (AmandeepArora) [reply C/services/south-plugin-interfaces/python/python_plugin_interface.cpp:82]: It is "`tf.__version__`"

**Comment by praveen-garg** [comment]
@AmandeepArora  could you please fix the name for `tests/system/plugins/filter/fledge-filter-imageblock` to `tests/system/plugins/filter/imageblock`? 
May be better to put this under tests/system/python/fledge/plugins/filter/imageblock

**Comment by praveen-garg** [comment]
@AmandeepArora  could you please fix the name for `tests/system/plugins/filter/fledge-filter-imageblock` to `tests/system/plugins/filter/imageblock`? 
May be better to put this under tests/system/python/fledge/plugins/filter/imageblock

**Comment by AmandeepArora** [comment]
> @AmandeepArora could you please fix the name for `tests/system/plugins/filter/fledge-filter-imageblock` to `tests/system/plugins/filter/imageblock`? May be better to put this under tests/system/python/fledge/plugins/filter/imageblock

Done

**Review by pintomax — DISMISSED** [review]

**Review by pintomax — APPROVED** [review]

**Review by AmandeepArora — APPROVED** [review]

**Review by pintomax — APPROVED** [review]

**Review by AmandeepArora — APPROVED** [review]

**Comment by YashTatkondawar** [comment]
As we have removed a couple of default north plugins(OCS and pi_server), we need to modify the API test https://github.com/fledge-iot/fledge/blob/develop/tests/system/python/api/test_plugin_discovery.py accordingly.

**Review by pintomax — DISMISSED** [review]

**Review by AmandeepArora — APPROVED** [review]

**Review by pintomax — DISMISSED** [review]

**Review by pintomax — DISMISSED** [review]

**Review by AmandeepArora — APPROVED** [review]

**Review by pintomax — APPROVED** [review]

**Pull request merged**